proposals/debug-id.md

```diff
@@ -0,0 +1,237 @@
+
+## Terms
+
+In the context of this document:
+
+- **Source Map:** Refers to a non-indexed, standard source map.
+- **Transpiled File:** Refers to a transpiled (potentially minified) JavaScript file.
+- **Debug ID:** Refers to a UUID as described in this document.
+
+## Debug IDs
+
+Debug IDs are globally unique identifiers for build artifacts.  They are
+specified to be UUIDs.  In the context of this proposal, they are represented in
+hexadecimal characters.  When comparing debug IDs they must be normalized.  This
+means that `85314830-023f-4cf1-a267-535f4e37bb17` and
+`85314830023F4CF1A267535F4E37BB17` are equivalent but the former representation
+is the canonical format.
+
+The way a debug ID is generated is specific to the toolchain and no requirements
+are placed on it.  It is however recommended to generate deterministic debug IDs
+(UUID v3 or v5) so that rebuilding the same artifacts yields stable IDs.
+
+Debug IDs are embedded in both source maps and transpiled files, allowing a
+bidirectional mapping between them. The linking of source maps and transpiled
+files via HTTP headers is explicitly not desired.  A file identified by a Debug
+ID must have that Debug ID embedded to ensure the file is self-identifying.
+
+### Debug IDs in Source Maps
+
+We propose adding a `debugId` property to the source map at the top level of
+the source map object.  This property should be a string value representing
+the Debug ID in hexadecimal characters, preferably in the canonical UUID
+format:
+
+```json
+{
+  "version": 3,
+  "file": "app.min.js",
+  "debugId": "85314830-023f-4cf1-a267-535f4e37bb17",
+  "sources": [...],
+  "sourcesContent": [...],
+  "mappings": "..."
+}
+```
+
+### Debug IDs in JavaScript Artifacts
+
+Transpiled JavaScript files containing a Debug ID must embed the ID near the end
+of the source, ideally on the last line, in the format `//# debugId=<DEBUG_ID>`:
+
+```javascript
+//# debugId=85314830-023f-4cf1-a267-535f4e37bb17
+```
+
+If the special `//# sourceMappingURL=` comment already exists in the file, it is
+recommended to place the `debugId` comment in the line above to maintain
+compatibility with existing tools.  Because the last line already has meaning in
+the existing specification for the `sourceMappingURL` comment, tools are
+required to examine the last 5 lines to discover the Debug ID.
+
+## JavaScript API for Debug ID Resolution
+
+Today `error.stack` in most runtimes only returns the URLs of the files referenced
+by the stack trace.  For Debug IDs to be useful, a solution would need to be added
+to enable mapping of JavaScript file URLs to Debug IDs.
+
+The strawman proposal is to add the Debug ID in two locations:
+
+* `import.meta.debugId`: a new property that should return the debug ID as UUID
+  of the current module if has one
+* `System.getDebugIdForUrl(url)` looks up the debug ID for a given script file by
+  URL that has already been loaded by the browser in the current context.
```

Comment:
I believe this might be the most controversial point of this proposal.

While a lot of tools use the `System` global, I believe no "official" spec is using it?

Also, "URL that has already been loaded by the browser" might need a bit more clarification, for example:
- It should return the debug-id of scripts "that are current loaded" into the engine.
- The URL should match the one that was used to load the script, and should also match the ones pretty-printed in `Error.stack` output, `import.meta.url`, etc.
- Specifically, it should have the same query-string and fragment.
- It should not do any network IO
- On mismatch, should it return `undefined` or rather throw a `ReferenceError` or any other kind of error?
- On mismatch, should it give an explanation of the reason? Is the file not loaded at all? Do the querystring parameters not match? etc…